app/jobs/delete_user_job.rb

```diff
@@ -4,13 +4,36 @@ class DeleteUserJob < ApplicationJob
 
   def perform(user:)
     email = user.email
-    user.destroy!
-  rescue ActiveRecord::RecordNotDestroyed, ActiveRecord::NotNullViolation, ActiveRecord::DeleteRestrictionError => e
-    # Catch the exception so we can log it, otherwise using `destroy` would give
-    # us no hint as to why the deletion failed.
-    Rails.error.report(e, context: { user:, email: }, handled: true)
-    Mailer.deletion_failed(email).deliver_later
-  else
-    Mailer.deletion_complete(email).deliver_later
+    return if user.reload.deleted_at?
+
+    begin
+      user.transaction do
+        user.yank_gems
+        user.api_keys.expire_all!
+        user.class.reflect_on_all_associations.each do |association|
+          next if association.name == :api_keys
+          case association.options[:dependent]
+          when :destroy, :destroy_async
+            user.association(association.name).handle_dependency
+          end
+        end
```

Comment:
I find it a little strange to manage the associations ourselves like this instead of explicitly deleting what should be deleted here.

An argument in favor of the approach in this PR is that you _could_ delete a user for real and it would cascade the delete correctly because you have all the dependent options set. But, do we ever do that?

I'm wondering if leaving the path open for a hard delete makes this unnecessarily complex. Reaching into active record reflections and handling it ourselves doesn't seem very forward compatible, and it seems to split the logic about what is actually happening here.

Reply:
let me know if you prefer this?